File: backend/src/main/java/com/cvmobile/cv/domain/model/Education.java
package com.cvmobile.cv.domain.model;

import java.time.LocalDate;

/**
 * Formation figurant sur un CV (valeur de domaine immuable).
 *
 * <p>Tous les champs sont facultatifs pour permettre l'edition partielle.
 *
 * @param id            identifiant technique, {@code null} tant que non persiste
 * @param etablissement nom de l'etablissement, ou {@code null}
 * @param diplome       intitule du diplome, ou {@code null}
 * @param domaine       domaine d'etude, ou {@code null}
 * @param dateDebut     debut de la periode, ou {@code null}
 * @param dateFin       fin de la periode, ou {@code null}
 * @param description   details, ou {@code null}
 */
public record Education(
        Long id,
        String etablissement,
        String diplome,
        String domaine,
        LocalDate dateDebut,
        LocalDate dateFin,
        String description) {

    /**
     * Construit une formation en normalisant les champs texte.
     */
    public static Education of(
            Long id,
            String etablissement,
            String diplome,
            String domaine,
            LocalDate dateDebut,
            LocalDate dateFin,
            String description) {
        return new Education(
                id,
                DomainText.normalize(etablissement),
                DomainText.normalize(diplome),
                DomainText.normalize(domaine),
                dateDebut,
                dateFin,
                DomainText.normalize(description));
    }

    /**
     * Derive une copie avec une nouvelle description (adaptation IA des
     * variantes), sans muter l'original.
     */
    public Education withDescription(String newDescription) {
        return new Education(
                id, etablissement, diplome, domaine, dateDebut, dateFin,
                DomainText.normalize(newDescription));
    }
}

File: mobile/lib/features/cv/domain/entities/education.dart
import 'copy_with_sentinel.dart';

/// Formation d'un CV, entite de domaine immuable et pure.
final class Education {
  final int? id;
  final String? etablissement;
  final String? diplome;
  final String? domaine;
  final DateTime? dateDebut;
  final DateTime? dateFin;
  final String? description;

  const Education({
    this.id,
    this.etablissement,
    this.diplome,
    this.domaine,
    this.dateDebut,
    this.dateFin,
    this.description,
  });

  Education copyWith({
    Object? id = unsetSentinel,
    Object? etablissement = unsetSentinel,
    Object? diplome = unsetSentinel,
    Object? domaine = unsetSentinel,
    Object? dateDebut = unsetSentinel,
    Object? dateFin = unsetSentinel,
    Object? description = unsetSentinel,
  }) {
    return Education(
      id: identical(id, unsetSentinel) ? this.id : id as int?,
      etablissement: identical(etablissement, unsetSentinel)
          ? this.etablissement
          : etablissement as String?,
      diplome:
          identical(diplome, unsetSentinel) ? this.diplome : diplome as String?,
      domaine:
          identical(domaine, unsetSentinel) ? this.domaine : domaine as String?,
      dateDebut: identical(dateDebut, unsetSentinel)
          ? this.dateDebut
          : dateDebut as DateTime?,
      dateFin: identical(dateFin, unsetSentinel)
          ? this.dateFin
          : dateFin as DateTime?,
      description: identical(description, unsetSentinel)
          ? this.description
          : description as String?,
    );
  }

  @override
  bool operator ==(Object other) =>
      other is Education &&
      other.id == id &&
      other.etablissement == etablissement &&
      other.diplome == diplome &&
      other.domaine == domaine &&
      other.dateDebut == dateDebut &&
      other.dateFin == dateFin &&
      other.description == description;

  @override
  int get hashCode => Object.hash(
        id,
        etablissement,
        diplome,
        domaine,
        dateDebut,
        dateFin,
        description,
      );
}
